src/Content/Item.php

```diff
@@ -756,4 +788,237 @@ public function addShareLink(string $body, int $quote_uri_id): string
+
+		if (!empty($attachment_img_src)) {
+			$attachment['images'] = [
+				0 => [
+					'src'    => $attachment_img_src,
+					'width'  => $attachment_img_width,
+					'height' => $attachment_img_height
+				]
+			];
+		} else {
+			unset($attachment['images']);
+		}
+
+		return "\n" . PageInfo::getFooterFromData($attachment);
+	}
+
+	public function addCategories(array $post, string $category): array
+	{
+		if (!empty($post['file'])) {
+			// get the "fileas" tags for this post
+			$filedas = FileTag::fileToArray($post['file']);
+		}
+
+		$list_array = explode(',', trim($category));
+		$post['file'] = FileTag::arrayToFile($list_array, 'category');
+
+		if (!empty($filedas) && is_array($filedas)) {
+			// append the fileas stuff to the new categories list
+			$post['file'] .= FileTag::arrayToFile($filedas);
+		}
+		return $post;
+	}
+
+	public function getACL(array $post, array $toplevel_item, array $request): array
+	{
+		// If this is a comment, set the permissions from the parent.
+		if ($toplevel_item) {
+			$post['allow_cid'] = $toplevel_item['allow_cid'] ?? '';
+			$post['allow_gid'] = $toplevel_item['allow_gid'] ?? '';
+			$post['deny_cid']  = $toplevel_item['deny_cid'] ?? '';
+			$post['deny_gid']  = $toplevel_item['deny_gid'] ?? '';
+			$post['private']   = $toplevel_item['private'];
+			return $post;
+		}
+
+		$user = User::getById($post['uid'], ['allow_cid', 'allow_gid', 'deny_cid', 'deny_gid']);
+		if (!DBA::isResult($user)) {
+			throw new HTTPException\NotFoundException($this->l10n->t('Unable to locate original post.'));
+		}
+
+		$post['allow_cid'] = isset($request['contact_allow']) ? $this->aclFormatter->toString($request['contact_allow']) : $user['allow_cid'] ?? '';
+		$post['allow_gid'] = isset($request['group_allow'])   ? $this->aclFormatter->toString($request['group_allow'])   : $user['allow_gid'] ?? '';
+		$post['deny_cid']  = isset($request['contact_deny'])  ? $this->aclFormatter->toString($request['contact_deny'])  : $user['deny_cid']  ?? '';
+		$post['deny_gid']  = isset($request['group_deny'])    ? $this->aclFormatter->toString($request['group_deny'])    : $user['deny_gid']  ?? '';
+
+		$visibility = $request['visibility'] ?? '';
+		if ($visibility === 'public') {
+			// The ACL selector introduced in version 2019.12 sends ACL input data even when the Public visibility is selected
+			$post['allow_cid'] = $post['allow_gid'] = $post['deny_cid'] = $post['deny_gid'] = '';
+		} else if ($visibility === 'custom') {
+			// Since we know from the visibility parameter the item should be private, we have to prevent the empty ACL
+			// case that would make it public. So we always append the author's contact id to the allowed contacts.
+			// See https://github.com/friendica/friendica/issues/9672
+			$post['allow_cid'] .= $this->aclFormatter->toString(Contact::getPublicIdByUserId($post['uid']));
+		}
+
+		if (strlen($post['allow_gid']) || strlen($post['allow_cid']) || strlen($post['deny_gid']) || strlen($post['deny_cid'])) {
```

Comment:
Superfluous function calls.
```suggestion
		if ($post['allow_gid'] || $post['allow_cid'] || $post['deny_gid'] || $post['deny_cid']) {
```